core/imt/test/testRTaskArena.cxx

```diff
@@ -141,4 +142,92 @@ TEST(RTaskArena, InterleaveAndNest)
    EXPECT_EQ(ROOT::Internal::RTaskArenaWrapper::TaskArenaSize(), 0u);
 }
 
+
+// Acquire pointers to ROOT's task arena from many threads in parallel.
+// To create more chaos, half of the threads will immediately try to get the pointer,
+// while the other half waits for a condition variable. When the first half
+// is done, the corresponding shared pointers to the task arena are destroyed
+// (and hence the task arena). Then, using a notify_all(), the other half of
+// threads will wake up and race to create a task arena again.
+TEST(RTaskArena, ThreadSafety) {
+   constexpr unsigned int nThreads = 500;
+   using TaskArenaPtr_t = std::shared_ptr<ROOT::Internal::RTaskArenaWrapper>;
+   bool firstHalfDone = false;
+   std::condition_variable cv;
+   std::mutex m;
+
+   std::vector<TaskArenaPtr_t> arenas1(nThreads/2);
+   std::vector<TaskArenaPtr_t> arenas2(nThreads/2);
+   std::vector<std::thread> threads;
+   for (unsigned int i=0; i < nThreads; ++i) {
+      threads.emplace_back([&,i]{
+         if (i < nThreads/2) {
+            arenas1.at(i) = ROOT::Internal::GetGlobalTaskArena();
+         } else {
+            bool canContinue = false;
+            do {
+               std::unique_lock<std::mutex> lock(m);
+               canContinue = cv.wait_for(lock, std::chrono::microseconds(10), [=](){return firstHalfDone;});
+            } while (!canContinue);
+            arenas2.at(i % arenas2.size()) = ROOT::Internal::GetGlobalTaskArena();
+         }
+      });
+   }
+
+   std::for_each(threads.begin(), threads.end()-nThreads/2, [](std::thread& thr){thr.join();});
+   const ROOT::Internal::RTaskArenaWrapper* const ptrFirstHalf = arenas1.front().get();
+   EXPECT_TRUE(std::all_of(arenas1.begin(), arenas1.end(), [ptrFirstHalf](const TaskArenaPtr_t& ptr){
+      return ptr.get() == ptrFirstHalf;
+   }));
+
+   // Destroys ROOT's thread pool
+   arenas1.clear();
+   {
+      std::unique_lock<std::mutex> lock(m);
+      firstHalfDone = true;
+      cv.notify_all();
+   }
```

Comment:
where is ROOT's thread pool destroyed here?

Reply:
Missing headers :+1:.

When the shared_ptrs to ROOT's task arena are `cleared()` from the vector, the arena disappears, because it's a function-local weak_ptr. It doesn't necessarily destroy the thread pool.

Reply:
Funny that mac doesn't miss those headers ...